solr/core/src/java/org/apache/solr/search/facet/FacetRangeProcessor.java

```diff
@@ -539,7 +617,116 @@ private  SimpleOrderedMap getRangeCountsIndexed() throws IOException {
+    if (hasSubFacets) {
+      filters[slot] = rangeQ;
+      intersections[slot] = intersection;  // save for later  // TODO: only save if number of slots is small enough?
+    }
+    int num = collect(intersection, slot, slotNum -> { return new SlotAcc.SlotContext(rangeQ); });
+    countAcc.incrementCount(slot, num); // TODO: roll this into collect()
+  }
+
+  private void dvRangeStats(List<Range> rangeList, List<Range> otherList, boolean hasSubFacets) throws IOException {
+    List<Range> allRanges = new ArrayList<>(rangeList.size() + otherList.size());
+    allRanges.addAll(rangeList); // first rangeList items are sorted by low
+    allRanges.addAll(otherList);
+    
+    final FieldType ft = sf.getType();
+    final String fieldName = sf.getName();
+    final NumberType numericType = ft.getNumberType();
+    if (numericType == null) {
+      throw new IllegalStateException();
+    }
+
+    final IndexReader indexReader = fcontext.searcher.getIndexReader();
+    final int maxDoc = indexReader.maxDoc();
+    final Iterator<LeafReaderContext> ctxIt = indexReader.leaves().iterator();
+    LeafReaderContext ctx = null;
+    NumericDocValues longs = null;
+    DocSetBuilder[] setBuilders = hasSubFacets ? new DocSetBuilder[intersections.length] : null;
+    
+    for (DocIterator docsIt = fcontext.base.iterator(); docsIt.hasNext(); ) {
+      final int doc = docsIt.nextDoc();
+      if (ctx == null || doc >= ctx.docBase + ctx.reader().maxDoc()) {
+        do {
+          ctx = ctxIt.next();
+        } while (ctx == null || doc >= ctx.docBase + ctx.reader().maxDoc());
+        assert doc >= ctx.docBase;
+        switch (numericType) {
+          case LONG:
+          case DATE:
+          case INTEGER:
+            longs = DocValues.getNumeric(ctx.reader(), fieldName);
+            break;
+          case FLOAT:
+            longs = new FilterNumericDocValues(DocValues.getNumeric(ctx.reader(), fieldName)) {
+              @Override
+              public long longValue() throws IOException {
+                return NumericUtils.sortableFloatBits((int) super.longValue());
+              }
+            };
+            break;
+          case DOUBLE:
+            longs = new FilterNumericDocValues(DocValues.getNumeric(ctx.reader(), fieldName)) {
+              @Override
+              public long longValue() throws IOException {
+                return NumericUtils.sortableDoubleBits(super.longValue());
+              }
+            };
+            break;
+          default:
+            throw new AssertionError();
+        }
+      }
+      final int localDoc = doc - ctx.docBase;
+      int valuesDocID = longs.docID();
+      if (valuesDocID < localDoc) {
+        valuesDocID = longs.advance(localDoc);
+      }
+      if (valuesDocID == localDoc) {
+        long value = longs.longValue();
+        // TODO: should we binary search first, or put in a tree set? gets weird with overlapping ranges
+        for (int idx = 0; idx < allRanges.size();) {
```

Comment:
Looking up the interval is an O(n) operation now. Shall we use an interval tree and make it an O(log n)? 

In the case of overlapping ranges, there are regular scenarios where the other buckets will be consecutive to the bucket we arrive at using binary search/interval tree's find. In non-regular cases ("other" parameter is called with non "none" value, or "include" is called with "all"), current approach seems reasonable. Reference: https://solr.apache.org/guide/8_8/json-facet-api.html#range-facet-parameters

Reply:
Yeah, good call on interval tree - let me see if there is something to use for that. I do intend to update some of our APIs to use `range` facets with the `ranges` param with possibly quite a few overlapping ranges (for daily, weekly, monthly average trends).

FWIW the params version of range faceting seemed to work much much better still with this impl - linear, but with skipping once the data point is lower than the start point. Would love to improve though!

Reply:
There isn't a great interval tree option available outside of `sun.jvm.hotspot.utilities` which feels strange to use - but I guess I'm going to try and update to it and see what it looks like.

As an aside, I realized I should probably try and use long primitives to avoid autoboxing. Going to update for that as well.

Reply:
Eh, nevermind, though it was available, but I'd have to add as a dep. Maybe something to consider as part of upstream and internal testing? Since the current formulation follows the params based impl I'll just leave as linear with minor optimizations for now.